File: frontend/src/services/websocketService.ts
/**
 * WebSocket Service
 * Real-time communication using Socket.IO
 */

import { io, Socket } from 'socket.io-client'
import { toast } from 'react-toastify'

export interface Notification {
  id: string
  user_id: string
  type: string
  title: string
  message: string
  metadata: Record<string, any>
  is_read: boolean
  read_at: string | null
  entity_type: string | null
  entity_id: string | null
  priority: string
  expires_at: string | null
  created_date: string
  updated_date: string
}

type NotificationCallback = (notification: Notification) => void
type SystemMessageCallback = (message: any) => void
type RealTimeUpdateCallback = (update: any) => void

class WebSocketService {
  private socket: Socket | null = null
  private notificationCallbacks: NotificationCallback[] = []
  private systemMessageCallbacks: SystemMessageCallback[] = []
  private realTimeUpdateCallbacks: Map<string, RealTimeUpdateCallback[]> = new Map()
  private reconnectAttempts = 0
  private maxReconnectAttempts = 5

  /**
   * Connect to WebSocket server
   */
  connect(token: string, userId: string, tenantId: string): void {
    if (this.socket?.connected) {
      console.log('WebSocket already connected')
      return
    }

    const wsUrl = import.meta.env.VITE_WS_URL || 'http://localhost:8000/ws'

    this.socket = io(wsUrl, {
      auth: {
        token,
      },
      transports: ['websocket', 'polling'],
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      reconnectionAttempts: this.maxReconnectAttempts,
    })

    this.setupEventHandlers(userId, tenantId)
  }

  /**
   * Setup Socket.IO event handlers
   */
  private setupEventHandlers(userId: string, tenantId: string): void {
    if (!this.socket) return

    // Connection events
    this.socket.on('connect', () => {
      console.log('WebSocket connected')
      this.reconnectAttempts = 0

      // Authenticate after connection
      this.socket?.emit('authenticate', {
        user_id: userId,
        tenant_id: tenantId,
      })

      toast.success('Real-time connection established', { autoClose: 2000 })
    })

    this.socket.on('authenticated', (data) => {
      console.log('WebSocket authenticated:', data)
    })

    this.socket.on('disconnect', (reason) => {
      console.log('WebSocket disconnected:', reason)
      toast.warning('Real-time connection lost', { autoClose: 3000 })
    })

    this.socket.on('connect_error', (error) => {
      console.error('WebSocket connection error:', error)
      this.reconnectAttempts++

      if (this.reconnectAttempts >= this.maxReconnectAttempts) {
        toast.error('Failed to establish real-time connection', { autoClose: 5000 })
      }
    })

    this.socket.on('reconnect', (attemptNumber) => {
      console.log('WebSocket reconnected after', attemptNumber, 'attempts')
      this.reconnectAttempts = 0
      toast.success('Real-time connection restored', { autoClose: 2000 })
    })

    // Notification event
    this.socket.on('notification', (notification: Notification) => {
      console.log('Received notification:', notification)

      // Call all registered callbacks
      this.notificationCallbacks.forEach((callback) => {
        try {
          callback(notification)
        } catch (error) {
          console.error('Error in notification callback:', error)
        }
      })

      // Show toast notification
      const toastOptions = {
        autoClose: 5000,
      }

      switch (notification.type) {
        case 'success':
          toast.success(notification.message, toastOptions)
          break
        case 'warning':
          toast.warning(notification.message, toastOptions)
          break
        case 'error':
          toast.error(notification.message, toastOptions)
          break
        default:
          toast.info(notification.message, toastOptions)
      }
    })

    // System message event
    this.socket.on('system_message', (message: any) => {
      console.log('Received system message:', message)

      this.systemMessageCallbacks.forEach((callback) => {
        try {
          callback(message)
        } catch (error) {
          console.error('Error in system message callback:', error)
        }
      })

      // Show toast for system messages
      const toastMessage = message.message || 'System notification'
      switch (message.level) {
        case 'warning':
          toast.warning(toastMessage, { autoClose: 10000 })
          break
        case 'error':
          toast.error(toastMessage, { autoClose: 10000 })
          break
        default:
          toast.info(toastMessage, { autoClose: 5000 })
      }
    })

    // Real-time update event
    this.socket.on('realtime_update', (update: any) => {
      console.log('Received real-time update:', update)

      const callbacks = this.realTimeUpdateCallbacks.get(update.type) || []
      callbacks.forEach((callback) => {
        try {
          callback(update.data)
        } catch (error) {
          console.error('Error in real-time update callback:', error)
        }
      })
    })

    // Subscription events
    this.socket.on('subscribed', (data) => {
      console.log('Subscribed to channel:', data.channel)
    })

    this.socket.on('unsubscribed', (data) => {
      console.log('Unsubscribed from channel:', data.channel)
    })

    // Error event
    this.socket.on('error', (error) => {
      console.error('WebSocket error:', error)
      toast.error(error.message || 'WebSocket error occurred', { autoClose: 5000 })
    })
  }

  /**
   * Disconnect from WebSocket server
   */
  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect()
      this.socket = null
      console.log('WebSocket disconnected manually')
    }
  }

  /**
   * Subscribe to a channel
   */
  subscribe(channel: string): void {
    if (this.socket?.connected) {
      this.socket.emit('subscribe', { channel })
    }
  }

  /**
   * Unsubscribe from a channel
   */
  unsubscribe(channel: string): void {
    if (this.socket?.connected) {
      this.socket.emit('unsubscribe', { channel })
    }
  }

  /**
   * Register a notification callback
   */
  onNotification(callback: NotificationCallback): () => void {
    this.notificationCallbacks.push(callback)

    // Return unsubscribe function
    return () => {
      const index = this.notificationCallbacks.indexOf(callback)
      if (index > -1) {
        this.notificationCallbacks.splice(index, 1)
      }
    }
  }

  /**
   * Register a system message callback
   */
  onSystemMessage(callback: SystemMessageCallback): () => void {
    this.systemMessageCallbacks.push(callback)

    return () => {
      const index = this.systemMessageCallbacks.indexOf(callback)
      if (index > -1) {
        this.systemMessageCallbacks.splice(index, 1)
      }
    }
  }

  /**
   * Register a real-time update callback
   */
  onRealTimeUpdate(eventType: string, callback: RealTimeUpdateCallback): () => void {
    if (!this.realTimeUpdateCallbacks.has(eventType)) {
      this.realTimeUpdateCallbacks.set(eventType, [])
    }

    this.realTimeUpdateCallbacks.get(eventType)!.push(callback)

    return () => {
      const callbacks = this.realTimeUpdateCallbacks.get(eventType)
      if (callbacks) {
        const index = callbacks.indexOf(callback)
        if (index > -1) {
          callbacks.splice(index, 1)
        }
      }
    }
  }

  /**
   * Check if WebSocket is connected
   */
  isConnected(): boolean {
    return this.socket?.connected || false
  }

  /**
   * Mark notification as read
   */
  markNotificationAsRead(notificationId: string): void {
    if (this.socket?.connected) {
      this.socket.emit('mark_notification_read', { notification_id: notificationId })
    }
  }
}

// Singleton instance
const websocketService = new WebSocketService()

export default websocketService

File: backend/app/services/websocket_service.py
"""
WebSocket Service
Real-time communication using Socket.IO
"""

import socketio
import logging
from typing import Dict, Set, Optional, Any
from uuid import UUID

logger = logging.getLogger(__name__)


class WebSocketService:
    """WebSocket service for real-time notifications"""

    def __init__(self):
        """Initialize Socket.IO server"""
        self.sio = socketio.AsyncServer(
            async_mode='asgi',
            cors_allowed_origins='*',
            logger=True,
            engineio_logger=True
        )

        # Track connected users: {user_id: set(session_ids)}
        self.user_sessions: Dict[str, Set[str]] = {}

        # Track session metadata: {session_id: {"user_id": str, "tenant_id": str}}
        self.session_metadata: Dict[str, Dict[str, str]] = {}

        # Setup event handlers
        self._setup_handlers()

    def _setup_handlers(self):
        """Setup Socket.IO event handlers"""

        @self.sio.event
        async def connect(sid, environ, auth):
            """Handle client connection"""
            logger.info(f"Client connecting: {sid}")
            logger.info(f"Auth data: {auth}")

            # Allow connection initially
            # Authentication will be verified in the 'authenticate' event
            logger.info(f"Client {sid} connected successfully (pending authentication)")
            return True

        @self.sio.event
        async def disconnect(sid):
            """Handle client disconnection"""
            logger.info(f"Client disconnected: {sid}")

            # Remove session from tracking
            if sid in self.session_metadata:
                metadata = self.session_metadata[sid]
                user_id = metadata.get("user_id")

                if user_id and user_id in self.user_sessions:
                    self.user_sessions[user_id].discard(sid)
                    if not self.user_sessions[user_id]:
                        del self.user_sessions[user_id]

                del self.session_metadata[sid]

        @self.sio.event
        async def authenticate(sid, data):
            """Authenticate user after connection"""
            user_id = data.get("user_id")
            tenant_id = data.get("tenant_id")

            if not user_id or not tenant_id:
                await self.sio.emit('error', {'message': 'Invalid authentication data'}, room=sid)
                return

            # Store session metadata
            self.session_metadata[sid] = {
                "user_id": str(user_id),
                "tenant_id": str(tenant_id)
            }

            # Track user sessions
            if user_id not in self.user_sessions:
                self.user_sessions[str(user_id)] = set()
            self.user_sessions[str(user_id)].add(sid)

            logger.info(f"User {user_id} authenticated on session {sid}")

            # Send authentication success
            await self.sio.emit('authenticated', {'status': 'success'}, room=sid)

        @self.sio.event
        async def subscribe(sid, data):
            """Subscribe to specific event channels"""
            channel = data.get("channel")
            if channel:
                await self.sio.enter_room(sid, channel)
                logger.info(f"Session {sid} subscribed to channel: {channel}")
                await self.sio.emit('subscribed', {'channel': channel}, room=sid)

        @self.sio.event
        async def unsubscribe(sid, data):
            """Unsubscribe from specific event channels"""
            channel = data.get("channel")
            if channel:
                await self.sio.leave_room(sid, channel)
                logger.info(f"Session {sid} unsubscribed from channel: {channel}")
                await self.sio.emit('unsubscribed', {'channel': channel}, room=sid)

        @self.sio.event
        async def mark_notification_read(sid, data):
            """Mark a notification as read"""
            notification_id = data.get("notification_id")
            if notification_id:
                # This would update the database
                # For now, just acknowledge
                await self.sio.emit('notification_marked_read', {
                    'notification_id': notification_id
                }, room=sid)

    async def send_notification_to_user(
        self,
        user_id: str,
        notification: Dict[str, Any]
    ):
        """
        Send notification to a specific user

        Args:
            user_id: User ID to send notification to
            notification: Notification data dict
        """
        user_id_str = str(user_id)

        if user_id_str not in self.user_sessions:
            logger.info(f"User {user_id_str} not connected, notification will be stored only")
            return

        # Send to all user's active sessions
        for session_id in self.user_sessions[user_id_str]:
            await self.sio.emit('notification', notification, room=session_id)

        logger.info(f"Notification sent to user {user_id_str} ({len(self.user_sessions[user_id_str])} sessions)")

    async def send_notification_to_tenant(
        self,
        tenant_id: str,
        notification: Dict[str, Any],
        exclude_user_id: Optional[str] = None
    ):
        """
        Send notification to all users in a tenant

        Args:
            tenant_id: Tenant ID
            notification: Notification data dict
            exclude_user_id: Optional user ID to exclude from broadcast
        """
        tenant_id_str = str(tenant_id)
        exclude_user_id_str = str(exclude_user_id) if exclude_user_id else None

        count = 0
        for session_id, metadata in self.session_metadata.items():
            if metadata.get("tenant_id") == tenant_id_str:
                # Skip excluded user
                if exclude_user_id_str and metadata.get("user_id") == exclude_user_id_str:
                    continue

                await self.sio.emit('notification', notification, room=session_id)
                count += 1

        logger.info(f"Notification broadcast to tenant {tenant_id_str} ({count} sessions)")

    async def broadcast_system_message(
        self,
        message: str,
        type: str = "info"
    ):
        """
        Broadcast system message to all connected users

        Args:
            message: Message to broadcast
            type: Message type (info, warning, error)
        """
        notification = {
            "type": "system",
            "message": message,
            "level": type
        }

        await self.sio.emit('system_message', notification)
        logger.info(f"System message broadcast: {message}")

    async def send_realtime_update(
        self,
        channel: str,
        event_type: str,
        data: Dict[str, Any]
    ):
        """
        Send real-time update to a specific channel

        Args:
            channel: Channel name (e.g., "circulation", "acquisitions")
            event_type: Event type (e.g., "checkout", "checkin")
            data: Event data
        """
        event_data = {
            "type": event_type,
            "data": data
        }

        await self.sio.emit('realtime_update', event_data, room=channel)
        logger.info(f"Real-time update sent to channel {channel}: {event_type}")

    def get_asgi_app(self):
        """Get ASGI app for integration with FastAPI"""
        return socketio.ASGIApp(
            self.sio,
            socketio_path='socket.io'
        )

    async def shutdown(self):
        """Shutdown WebSocket service"""
        logger.info("Shutting down WebSocket service")
        await self.sio.disconnect()


# Singleton instance
_websocket_service: Optional[WebSocketService] = None


def get_websocket_service() -> WebSocketService:
    """Get or create WebSocket service singleton"""
    global _websocket_service

    if _websocket_service is None:
        _websocket_service = WebSocketService()

    return _websocket_service


async def init_websocket():
    """Initialize WebSocket service"""
    service = get_websocket_service()
    logger.info("WebSocket service initialized")
    return service


async def close_websocket():
    """Close WebSocket service"""
    global _websocket_service

    if _websocket_service:
        await _websocket_service.shutdown()
        _websocket_service = None

    logger.info("WebSocket service closed")
